### PR TITLE
crimson: peering event processing fixes,  wait for async operations started during peering events

### PR DESCRIPTION
**Pull request by athanatos**
<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
crimson-rados/thrash --limit 10 sjust-2024-07-06_06:23:37-crimson-rados:thrash-wip-sjust-crimson-testing-2024-07-05-distro-default-smithi: one notable failure, looks good otherwise
- 7789460 FuturizedCollection crash during load_pogs, https://tracker.ceph.com/issues/66294

Full crimson-rados run pending (can't access pulpito for link right now)

**Comment by athanatos** [comment]
jenkins test make check

**Comment by athanatos** [comment]
https://tracker.ceph.com/issues/66316

**Comment by athanatos** [comment]
jenkins test api

**Comment by Svelar** [comment]
jenkins test make check arm64

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test api

File: src/crimson/osd/osd_operations/peering_event.cc [hunk 1]
```diff
@@ -102,12 +102,11 @@ seastar::future<> PeeringEvent<T>::with_pg(
        * now, but we'll want to remove that as well.
        * https://tracker.ceph.com/issues/66708
        */
-      return interruptor::async([this, pg] {
+      return interruptor::async([this, pg, &shard_services] {
 	pg->do_peering_event(evt, ctx);
+	complete_rctx(shard_services, pg).get();
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/osd_operations/peering_event.cc:107]: I'd rather leave it as is -- pg_advance_map does potentially multiple events before calling complete_rctx.

File: src/crimson/osd/osd_operations/peering_event.cc [hunk 2]
```diff
@@ -89,8 +89,22 @@ seastar::future<> PeeringEvent<T>::with_pg(
     }).then_interruptible([this, pg](auto) {
       return this->template enter_stage<interruptor>(peering_pp(*pg).process);
     }).then_interruptible([this, pg, &shard_services] {
-      return pg->do_peering_event(evt, ctx
-      ).then_interruptible([this] {
+      /* The DeleteSome event invokes PeeringListener::do_delete_work, which
+       * needs to return (without a future) the object to start with on the next
+       * call.  As a consequence, crimson's do_delete_work implementation needs
+       * to use get0() for the object listing.  To support that, we wrap
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/osd_operations/peering_event.cc:95]: Ah, and that's what do_delete_work actually does -- fixing.

**Comment by Matan-B** [comment]
rescheduled: https://pulpito.ceph.com/matan-2024-07-11_09:18:39-crimson-rados-wip-sjust-crimson-testing-2024-07-05-distro-crimson-smithi/

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
Looking at the bluestore failures, this one looks new:
https://pulpito.ceph.com/matan-2024-07-11_09:18:39-crimson-rados-wip-sjust-crimson-testing-2024-07-05-distro-crimson-smithi/7796240
osd0:
```
ceph-osd: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4726-gcdabdb32/rpm/el9/BUILD/ceph-19.0.0-4726-gcdabdb32/redhat-linux-build/boost/include/boost/smart_ptr/intrusive_ptr.hpp:201: T* boost::intrusive_ptr<T>::operator->() const [with T = crimson::osd::PG]: Assertion `px != 0' failed.
 5# boost::intrusive_ptr<crimson::osd::PG>::operator->() const in ceph-osd
 6# crimson::osd::InternalClientRequest::InternalClientRequest(boost::intrusive_ptr<crimson::osd::PG>) in ceph-osd
 7# crimson::osd::WatchTimeoutRequest::WatchTimeoutRequest(seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>) in ceph-osd
 8# auto crimson::OperationRegistryI::create_operation<crimson::osd::WatchTimeoutRequest, seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>&>(seastar::shared_ptr<crimson::osd::Watch>&&, boost::intrusive_ptr<crimson::osd::PG>&) in ceph-osd
 9# auto crimson::osd::PerShardState::start_operation<crimson::osd::WatchTimeoutRequest, seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>&>(seastar::shared_ptr<crimson::osd::Watch>&&, boost::intrusive_ptr<crimson::osd::PG>&) in ceph-osd
10# auto crimson::osd::ShardServices::start_operation<crimson::osd::WatchTimeoutRequest, seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>&>(seastar::shared_ptr<crimson::osd::Watch>&&, boost::intrusive_ptr<crimson::osd::PG>&) in ceph-osd
11# crimson::osd::Watch::do_watch_timeout() in ceph-osd
12# crimson::osd::Watch::Watch(crimson::osd::Watch::private_ctag_t, boost::intrusive_ptr<crimson::osd::ObjectContext>, watch_info_t const&, entity_name_t const&, boost::intrusive_ptr<crimson::osd::PG>)::{lambda()#1}::operator()() const in ceph-osd
13# seastar::noncopyable_function<void ()>::direct_vtable_for<crimson::osd::Watch::Watch(crimson::osd::Watch::private_ctag_t, boost::intrusive_ptr<crimson::osd::ObjectContext>, watch_info_t const&, entity_name_t const&, boost::intrusive_ptr<crimson::osd::PG>)::{lambda()#1}>::call(seastar::noncopyable_function<void ()> const*) in ceph-osd
```

Also this one, although Looks unrelated to me, what do you think?

https://pulpito.ceph.com/matan-2024-07-11_09:18:39-crimson-rados-wip-sjust-crimson-testing-2024-07-05-distro-crimson-smithi/7796255
0SD0:
```
0x60b000031370 is located 0 bytes inside of 104-byte region [0x60b000031370,0x60b0000313d8)
freed by thread T2 here:
    #0 0x7fbbb10e0ea0 in operator delete(void*, unsigned long) (/lib64/libasan.so.8+0xe0ea0) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
    #1 0x41e8f87 in std::__new_allocator<std::__detail::_Hash_node_base*>::deallocate(std::__detail::_Hash_node_base**, unsigned long) (/usr/bin/ceph-osd+0x41e8f87) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)

previously allocated by thread T0 here:
    #0 0x7fbbb10dfdf8 in operator new(unsigned long) (/lib64/libasan.so.8+0xdfdf8) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
    #1 0x41e9ab0 in std::__new_allocator<std::__detail::_Hash_node_base*>::allocate(unsigned long, void const*) (/usr/bin/ceph-osd+0x41e9ab0) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)

Thread T1 created by T0 here:
    #0 0x7fbbb1049395 in pthread_create (/lib64/libasan.so.8+0x49395) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
    #1 0xc47c625 in seastar::posix_thread::posix_thread(seastar::posix_thread::attr, std::function<void ()>) (/usr/bin/ceph-osd+0xc47c625) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)

Thread T2 created by T0 here:
    #0 0x7fbbb1049395 in pthread_create (/lib64/libasan.so.8+0x49395) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
    #1 0xc47c625 in seastar::posix_thread::posix_thread(seastar::posix_thread::attr, std::function<void ()>) (/usr/bin/ceph-osd+0xc47c625) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)

SUMMARY: AddressSanitizer: double-free (/lib64/libasan.so.8+0xe0ea0) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33) in operator delete(void*, unsigned long)
==32701==ABORTING
```

**Comment by athanatos** [comment]
I removed the Fixes: line from the last commit message as I saw an instance of the AsyncReserver crash with this PR merged.  I'd still like to go ahead and merge this as it's a useful cleanup and does seem to have greatly reduced the incidence rate, but I'm going to put together a followup.

**Comment by athanatos** [comment]
> Looking at the bluestore failures, this one looks new (may relate to #58463 because of InternalClientRequest edit): https://pulpito.ceph.com/matan-2024-07-11_09:18:39-crimson-rados-wip-sjust-crimson-testing-2024-07-05-distro-crimson-smithi/7796240 osd0:
> 
> ```
> ceph-osd: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-4726-gcdabdb32/rpm/el9/BUILD/ceph-19.0.0-4726-gcdabdb32/redhat-linux-build/boost/include/boost/smart_ptr/intrusive_ptr.hpp:201: T* boost::intrusive_ptr<T>::operator->() const [with T = crimson::osd::PG]: Assertion `px != 0' failed.
>  5# boost::intrusive_ptr<crimson::osd::PG>::operator->() const in ceph-osd
>  6# crimson::osd::InternalClientRequest::InternalClientRequest(boost::intrusive_ptr<crimson::osd::PG>) in ceph-osd
>  7# crimson::osd::WatchTimeoutRequest::WatchTimeoutRequest(seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>) in ceph-osd
>  8# auto crimson::OperationRegistryI::create_operation<crimson::osd::WatchTimeoutRequest, seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>&>(seastar::shared_ptr<crimson::osd::Watch>&&, boost::intrusive_ptr<crimson::osd::PG>&) in ceph-osd
>  9# auto crimson::osd::PerShardState::start_operation<crimson::osd::WatchTimeoutRequest, seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>&>(seastar::shared_ptr<crimson::osd::Watch>&&, boost::intrusive_ptr<crimson::osd::PG>&) in ceph-osd
> 10# auto crimson::osd::ShardServices::start_operation<crimson::osd::WatchTimeoutRequest, seastar::shared_ptr<crimson::osd::Watch>, boost::intrusive_ptr<crimson::osd::PG>&>(seastar::shared_ptr<crimson::osd::Watch>&&, boost::intrusive_ptr<crimson::osd::PG>&) in ceph-osd
> 11# crimson::osd::Watch::do_watch_timeout() in ceph-osd
> 12# crimson::osd::Watch::Watch(crimson::osd::Watch::private_ctag_t, boost::intrusive_ptr<crimson::osd::ObjectContext>, watch_info_t const&, entity_name_t const&, boost::intrusive_ptr<crimson::osd::PG>)::{lambda()#1}::operator()() const in ceph-osd
> 13# seastar::noncopyable_function<void ()>::direct_vtable_for<crimson::osd::Watch::Watch(crimson::osd::Watch::private_ctag_t, boost::intrusive_ptr<crimson::osd::ObjectContext>, watch_info_t const&, entity_name_t const&, boost::intrusive_ptr<crimson::osd::PG>)::{lambda()#1}>::call(seastar::noncopyable_function<void ()> const*) in ceph-osd
> ```
> 
> Same here: https://pulpito.ceph.com/matan-2024-07-11_09:18:39-crimson-rados-wip-sjust-crimson-testing-2024-07-05-distro-crimson-smithi/7796239 - osd0,2 and 3

Yeah, that one was a bug in the other PR, pushed a fix.

> 
> Also this one, although Looks unrelated to me, what do you think?
> 
> https://pulpito.ceph.com/matan-2024-07-11_09:18:39-crimson-rados-wip-sjust-crimson-testing-2024-07-05-distro-crimson-smithi/7796255 0SD0:
> 
> ```
> 0x60b000031370 is located 0 bytes inside of 104-byte region [0x60b000031370,0x60b0000313d8)
> freed by thread T2 here:
>     #0 0x7fbbb10e0ea0 in operator delete(void*, unsigned long) (/lib64/libasan.so.8+0xe0ea0) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
>     #1 0x41e8f87 in std::__new_allocator<std::__detail::_Hash_node_base*>::deallocate(std::__detail::_Hash_node_base**, unsigned long) (/usr/bin/ceph-osd+0x41e8f87) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)
> 
> previously allocated by thread T0 here:
>     #0 0x7fbbb10dfdf8 in operator new(unsigned long) (/lib64/libasan.so.8+0xdfdf8) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
>     #1 0x41e9ab0 in std::__new_allocator<std::__detail::_Hash_node_base*>::allocate(unsigned long, void const*) (/usr/bin/ceph-osd+0x41e9ab0) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)
> 
> Thread T1 created by T0 here:
>     #0 0x7fbbb1049395 in pthread_create (/lib64/libasan.so.8+0x49395) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
>     #1 0xc47c625 in seastar::posix_thread::posix_thread(seastar::posix_thread::attr, std::function<void ()>) (/usr/bin/ceph-osd+0xc47c625) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)
> 
> Thread T2 created by T0 here:
>     #0 0x7fbbb1049395 in pthread_create (/lib64/libasan.so.8+0x49395) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33)
>     #1 0xc47c625 in seastar::posix_thread::posix_thread(seastar::posix_thread::attr, std::function<void ()>) (/usr/bin/ceph-osd+0xc47c625) (BuildId: b9f4447a02ea9ec3eacc625832ebe799c1e05192)
> 
> SUMMARY: AddressSanitizer: double-free (/lib64/libasan.so.8+0xe0ea0) (BuildId: e72832baf1a219a1019b9ecbf8330cba69f7ad33) in operator delete(void*, unsigned long)
> ==32701==ABORTING
> ```

Yeah, that's a crash during AlienStore::open_collection as part of load_pgs -- https://tracker.ceph.com/issues/66294.  I'm testing a fix for that one now -- coll_map needs a mutex or we get undefined behavior as it's accessed from different reactor threads concurrently.

**Comment by Matan-B** [comment]
jenkins test api

**Comment by athanatos** [comment]
Modifying PR to address the crash I mentioned above -- let's hold off on merging it until I've got the new version tested.

**Comment by athanatos** [comment]
10 job rados/thrash run, no failures: https://pulpito.ceph.com/sjust-2024-07-23_08:42:01-crimson-rados:thrash-wip-sjust-crimson-testing-2024-07-22-distro-default-smithi/

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-07-24_07:30:51-crimson-rados-wip-sjust-crimson-testing-2024-07-22-distro-crimson-smithi/
Tests look great.
Only (bluestore's) rbd api failed (due to https://github.com/ceph/ceph/pull/57797) - unrelated.

**Comment by Matan-B** [comment]
jenkins test windows

**Review by Matan-B — APPROVED** [review]

**Pull request merged**